temporal-sdk/src/main/java/io/temporal/internal/worker/EagerActivitySlotsReservation.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (C) 2022 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ * Copyright (C) 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this material except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.internal.worker;
+
+import io.temporal.api.command.v1.Command;
+import io.temporal.api.command.v1.ScheduleActivityTaskCommandAttributes;
+import io.temporal.api.enums.v1.CommandType;
+import io.temporal.api.workflowservice.v1.PollActivityTaskQueueResponse;
+import io.temporal.api.workflowservice.v1.RespondWorkflowTaskCompletedRequest;
+import io.temporal.api.workflowservice.v1.RespondWorkflowTaskCompletedResponse;
+
+public class EagerActivitySlotsReservation implements AutoCloseable {
+  private final EagerActivityInjector eagerActivityInjector;
+  private int acquiredReservationCount = 0;
+
+  EagerActivitySlotsReservation(EagerActivityInjector eagerActivityInjector) {
+    this.eagerActivityInjector = eagerActivityInjector;
+  }
+
+  public RespondWorkflowTaskCompletedRequest applyToRequest(
+      RespondWorkflowTaskCompletedRequest request) {
+    RespondWorkflowTaskCompletedRequest.Builder mutableRequest = request.toBuilder();
+
+    for (int i = 0; i < mutableRequest.getCommandsCount(); i++) {
```

Comment:
I really don't like that we tinker with commands AFTER they got created. We don't do it with any other commands and we should avoid it here too. Why not move this into Activity StateMachine?